src/core/web-server.cc

```diff
@@ -530,6 +531,230 @@ class CDExecutor : public PCSX::WebExecutor {
     virtual ~CDExecutor() = default;
 };
 
+class StateExecutor : public PCSX::WebExecutor {
+    virtual bool match(PCSX::WebClient* client, const PCSX::UrlData& urldata) final {
+        return PCSX::StringsHelpers::startsWith(urldata.path, c_prefix);
+    }
+    virtual bool execute(PCSX::WebClient* client, PCSX::RequestData& request) final {
+        if (PCSX::g_gui == nullptr) {
+            client->write("HTTP/1.1 400 Bad Request\r\n\r\nSave states unavailable in CLI/no-UI mode.");
```

Comment:
We might want to move the savestate code to the general UI class instead of the GUI one, but that's for another PR.